### PR TITLE
Fixed issue #18976: Saving a question with subquestions deletes default answers

### PR DESCRIPTION
**Pull request by adamzammit**
Fixed issue #18976: Saving a question with subquestions deletes default answers
Dev: Delete missing subquestions from post, then add new subquestions, then update existing subquestions instead of deleting all and adding again.
Dev: Also base updates on id instead of a search by question code

### CODE REVIEW

File: application/controllers/QuestionAdministrationController.php [hunk 1]
```diff
@@ -2785,50 +2794,73 @@ function ($collector, $oDefaultValue) use ($oQuestion) {
...
      * @return void
      * @throws CHttpException
      */
-    private function storeSubquestions($question, $subquestionsArray)
+    private function deleteMissingSubquestions($question, $subquestionsArray)
+    {
+        foreach ($question->subquestions as $subquestion) {
+            if (!isset($subquestionsArray[$subquestion->qid])) {
+                if (!$subquestion->delete()) {
+                    throw (new LSUserException(500, gT("Could not delete subquestion")))
```
Comment (Shnoulle) [inline application/controllers/QuestionAdministrationController.php:2810]: The really throw error or add error after page display ?

File: application/controllers/QuestionAdministrationController.php [hunk 2]
```diff
@@ -473,11 +473,20 @@ public function actionSaveQuestionData()
             }
 
             if ($question->survey->active == 'N') {
-                // Clean subquestions before save.
-                $question->deleteAllSubquestions();
                 // If question type has subquestions, save them.
                 if ($question->questionType->subquestions > 0) {
-                    $this->storeSubquestions(
+                    // Delete subquestions that do not appear in submit
```
Comment (Shnoulle) [inline application/controllers/QuestionAdministrationController.php:478]: Why not order

1. deleteMissingSubquestions
2. updateSubquestions
3. addNewSubquestions

With updateSubquestions : checking if question exist (not only `new`) and same for addNewSubquestions ?

File: application/controllers/QuestionAdministrationController.php [hunk 3]
```diff
@@ -2852,54 +2885,50 @@ private function updateSubquestions($question, $subquestionsArray)
...
-                            'language' => $lang
-                        ]
-                    );
-                    if (empty($l10n)) {
-                        $l10n = new QuestionL10n();
+            if (substr($subquestionId, 0, 3) !== "new") {
+                foreach ($subquestionArray as $scaleId => $data) {
+                    $subquestion = Question::model()->findByPk($subquestionId);
+                    if (empty($subquestion)) {
+                        throw new Exception('Found no subquestion with code ' . $subquestionId);
```
Comment (Shnoulle) [inline application/controllers/QuestionAdministrationController.php:2892]: `Found no subquestion with id`

BUT : i don't think we must throw an error , maybe add a flashMessage + create the sub-question ?

File: application/controllers/QuestionAdministrationController.php [hunk 4]
```diff
@@ -2852,54 +2885,50 @@ private function updateSubquestions($question, $subquestionsArray)
...
-                    $l10n->language = $lang;
-                    $l10n->question = $questionText;
-                    if (!$l10n->save()) {
+                    $subquestion->sid        = $question->sid;
+                    $subquestion->gid        = $question->gid;
+                    $subquestion->parent_qid = $question->qid;
+                    $subquestion->question_order = $questionOrder;
+                    $questionOrder++;
+                    if (!isset($data['code'])) {
+                        throw new CHttpException(
```
Comment (Shnoulle) [inline application/controllers/QuestionAdministrationController.php:2900]: Came from previous version : OK.

But maybe take from current if not set ?

File: application/controllers/QuestionAdministrationController.php [hunk 5]
```diff
@@ -2785,50 +2794,73 @@ function ($collector, $oDefaultValue) use ($oQuestion) {
...
+            if (substr($subquestionId, 0, 3) === 'new') {
+                foreach ($subquestionArray as $scaleId => $data) {
+                    $subquestion = new Question();
+                    $subquestion->sid        = $question->sid;
+                    $subquestion->gid        = $question->gid;
+                    $subquestion->parent_qid = $question->qid;
+                    $subquestion->question_order = $questionOrder;
+                    $questionOrder++;
+                    if (!isset($data['code'])) {
+                        throw new CHttpException(
```
Comment (Shnoulle) [inline application/controllers/QuestionAdministrationController.php:2840]: Same than update but create a random one ?

**Review by Shnoulle — COMMENTED** [review]
Seems really cool !

Just some remark and potential improvment : i hate to loose data sent after update a forum, then (maybe except if debug=2) : add errors but try to save all values sent ?

**Comment by Shnoulle** [comment]
@gabrieljenik : a second view here ?

The fix by Adam is great, my comments are comments only.  You advice ?

**Review by Shnoulle — APPROVED** [review]

File: application/controllers/QuestionAdministrationController.php (hunk 2 above)
Comment (gabrieljenik) [reply application/controllers/QuestionAdministrationController.php:478]: Sorry, don't understand... what's the upside of the order change?

File: application/controllers/QuestionAdministrationController.php (hunk 1 above)
Comment (gabrieljenik) [reply application/controllers/QuestionAdministrationController.php:2810]: LSUserException has frieldy UI if I recall correctly.
Not sure if 500 is the right code.

File: application/controllers/QuestionAdministrationController.php [hunk 6]
```diff
@@ -2785,50 +2794,73 @@ function ($collector, $oDefaultValue) use ($oQuestion) {
...
-                $subquestion->refresh();
-                foreach ($data['subquestionl10n'] as $lang => $questionText) {
-                    $l10n = new QuestionL10n();
-                    $l10n->qid = $subquestion->qid;
-                    $l10n->language = $lang;
-                    $l10n->question = $questionText;
-                    if (!$l10n->save()) {
-                        throw (new LSUserException(500, gT("Could not save subquestion")))
-                            ->setDetailedErrorsFromModel($l10n);
+            if (substr($subquestionId, 0, 3) === 'new') {
```
Comment (gabrieljenik) [inline application/controllers/QuestionAdministrationController.php:2831]: This is very personal, but I like better this form of `if` when not having an else and inside a loop:
```
foreach() {
   if(!xxx) {
     continue;
   }
}
```
Flatter code.

Comment (gabrieljenik) [reply application/controllers/QuestionAdministrationController.php:2831]: Also, isn't a supposition all in `subquestionsArray` will be new?
I feel some behaviour should be moved out of this function as to have a single objetive.
Right now it has 2:
- distinguish what to add new and what not.
- add new.

File: application/controllers/QuestionAdministrationController.php (hunk 2 above)
Comment (Shnoulle) [reply application/controllers/QuestionAdministrationController.php:478]: after adding new subquestions : you have new qid related to this subqid.

Then you can not really check if question exist 

It's not a real issue , but seems more logical : deleted , update , add …

File: application/controllers/QuestionAdministrationController.php (hunk 6 above)
Comment (Shnoulle) [reply application/controllers/QuestionAdministrationController.php:2831]: :+1: here : really more clear !

File: application/controllers/QuestionAdministrationController.php (hunk 3 above)
Comment (Shnoulle) [reply application/controllers/QuestionAdministrationController.php:2892]: PS : must not happen

File: application/controllers/QuestionAdministrationController.php (hunk 6 above)
Comment (Shnoulle) [reply application/controllers/QuestionAdministrationController.php:2831]: > Also, isn't a supposition all in `subquestionsArray` will be new?

No : it's the reason of the issue : loosing default value on existing subquestion

**Comment by gabrieljenik** [comment]
Left a few comments through the code that could easily aply to other parts of this PR.

Also, on general basis, I would add 
- point the PR to dev
- take the oportunity to reorganize the code into a service/helper class (there is already quite some code movement, so why not putting it on a separate place)
- make new methods public, for easier unit testing.
- add unit testing

How this impacts RC code? Should that be adapted as well for coherence? (on anoher ticket or PR maybe?)

File: application/controllers/QuestionAdministrationController.php (hunk 6 above)
Comment (gabrieljenik) [reply application/controllers/QuestionAdministrationController.php:2831]: Still, I think this applies.

> I feel some behaviour should be moved out of this function as to have a single objetive.
> Right now it has 2:
> - distinguish what to add new and what not.
> - add new.

Comment (Shnoulle) [reply application/controllers/QuestionAdministrationController.php:2831]: > Still, I think this applies.
> 
> > I feel some behaviour should be moved out of this function as to have a single objetive.
> > Right now it has 2:
> > 
> > * distinguish what to add new and what not.
> > * add new.

Something like this : 

```
function addOrUpdateSubquestions($question, $subquestionsArray) {
    $questionOrder = 0;
    foreach ($subquestionsArray as $subquestionId => $subquestionArray ) {
        if (substr($subquestionId, 0, 3) === 'new') {
            $this->addNewSubquestion($question, $subquestionArray, $questionOrder);
        } else {
            $this->updateSubquestion($question, $subquestionId, $subquestionArray, $questionOrder);
        }
    }
}
```

?

Comment (gabrieljenik) [reply application/controllers/QuestionAdministrationController.php:2831]: Yes... something like that.
addOrUpdateSubquestions could be in the controller while updateSubquestion and addNewSubquestion on the service helper.

**Comment by adamzammit** [comment]
I agree with @Shnoulle  that this is a bug and should be resolved in master/5.x asap and other refactoring can be done in dev.

Do any changes need to be made for this to be accepted and merged as it is, as a bug fix?

**Comment by gabrieljenik** [comment]
> * take the oportunity to reorganize the code into a service/helper class (there is already quite some code movement, so why not putting it on a separate place)
> * make new methods public, for easier unit testing.

I would do as follows:
- take the oportunity to reorganize the code into a service/helper class
- make new methods public, for easier unit testing
- Review latest comments through the code regarding:
  +  "if" inside loops
  + purpose of functions

**Comment by adamzammit** [comment]
@gabrieljenik do you mean you will make these changes, or that you would want me to make them to accept this PR?

My opinion is making more structural changes to the code is probably something for dev not for a bug fix?

**Comment by Shnoulle** [comment]
> My opinion is making more structural changes to the code is probably something for dev not for a bug fix?

Same opinion here.

Else : reviewing the code : remove the double foreach can be a good idea.

```
function addOrUpdateSubquestions($question, $subquestionsArray) {
    $questionOrder = 0;
    foreach ($subquestionsArray as $subquestionId => $subquestionArray ) {
        if (substr($subquestionId, 0, 3) === 'new') {
            $this->addNewSubquestion($question, $subquestionArray, $questionOrder);
        } else {
            $this->updateSubquestion($question, $subquestionId, $subquestionArray, $questionOrder);
        }
    }
}
```

This remove the need of 

```
   if(!xxx) {
     continue;
   }
```


And we clearly **need** a unit test here.


Else : about throw error VS add error whe return : advice ?

**Comment by gabrieljenik** [comment]
> @gabrieljenik do you mean you will make these changes, or that you would want me to make them to accept this PR?

That those changes are, in my point of view, required for accepting the PR.
 I believe those are the standards applied to PRs we push.

> My opinion is making more structural changes to the code is probably something for dev not for a bug fix?

There are quite lot of changes already. I am just saying to move them into a different class.
That will make things more testeable. Also, as that will need to be done eventually, by doing it now we will minimize risk as we will take advantage of the focus being given now as well as the testing phase.

There is a balance in between doing the minimal approahc for fixing a bug and taking the opportunity to build long lasting code (refactoring a bit). On this case, I believe the change is already non-minimal, so we can go and do it long-lasting.

**Comment by gabrieljenik** [comment]
@olleharstedt Thoughts?

**Comment by olleharstedt** [comment]
Right, @kevin-foster-uk is working on some refactoring here already, his eyes might help to avoid huge merge conflicts moving forward.

**Comment by olleharstedt** [comment]
Let's wait until @kevin-foster-uk had a chance to inspect it, he's the one working on refactoring right now.

**Comment by kevin-foster-uk** [comment]
Hi @adamzammit, thanks for your effort on this. Sorry to do this again but this PR is too complicated to apply, since we have already done allot of refactoring on this code. 

I have created a simpler fix for the bug in this PR: https://github.com/LimeSurvey/LimeSurvey/pull/3357

This will be merged to master shortly.

**Comment by adamzammit** [comment]
thanks for this fix @kevin-foster-uk all good.

**Pull request closed without merging**